0.57 Improve reprocessor

FedericoCeratto opened the pull request:
Support Clickhouse

Commit:
0.57 Improve reprocessor

Commit:
Improve dryrun

Commit:
Do not generate UID twice. Improve debugging

Commit:
Handle broken measurements better

Commit:
Implement new trivial ID

Commit:
Switch to new trivial id

Commit:
Add typing

Commit:
Cleanup

Commit:
Cleanup

Commit:
Cleanup

Commit:
Minor cleanup

Commit:
Handle Meek

Commit:
Drop sshfeeder

hellais commented on af/fastpath/fastpath/core.py:
```diff
@@ -1463,7 +1464,7 @@ def flag_measurements_with_wrong_date(msm: dict, msmt_uid: str, scores: dict) ->
     try:
         recv_time = datetime.strptime(msmt_uid[:22], "%Y%m%d%H%M%S.%f_")
-        start_time = msm.get("measurement_start_time")
+        start_time = msm.get("measurement_start_time") or ""
```
I guess the gotcha here is that we have using the `or ""` to convert a `None` `measurement_start_time` into a string so that the following call will fail with a `ValueError` instead of a `TypeError`.

FedericoCeratto replied:
They are equally good at raising an exception but avoiding an optional None makes mypy happier

hellais commented on af/fastpath/fastpath/reprocessor.py:
```diff
@@ -163,55 +199,69 @@ def score_measurement_and_upsert_fastpath(msm, msmt_uid, do_update: bool) -> Non
+        cc = msm.get("probe_cc").upper()
+        desc = f"{msmt_uid} {tn} {cc} {rid} {input_}"
+    except Exception as e:
+        log.info(f"Ignoring broken measurement")
```
Can we add some additional logging information to identify which measurements are broken?

I think that once we do a run it would be useful to take a look at the broken ones and understand why we failed to load them.

I would say it's probably useful to record:
* Where it came from (the can or jsonl)
* The measurement_uid
* Increment a metric in statsd

FedericoCeratto replied:
Added "{desc}" and metric

hellais commented on af/fastpath/fastpath/normalize.py:
```diff
@@ -575,21 +576,24 @@ def iter_yaml_msmt_normalized(data, bucket_tstamp: str, report_fn: str):
         try:
-            yield normalize_entry(entry, bucket_tstamp, report_fn, esha)
+            d = normalize_entry(entry, bucket_tstamp, report_fn, esha_d)
+            msmt_uid = trivial_id(raw_entry, d)
```
Do we care to concatenate the entry header when computing the trivial_id?

FedericoCeratto replied:
What's the benefit?

hellais replied:
I guess that morally a YAML measurement is the header + the entry, but yeah you are probably right that there isn't any real benefit to doing so.

hellais replied:
We discussed this and it's probably not so helpful to have. Closing.

hellais commented on af/fastpath/fastpath/s3feeder.py:
```diff
@@ -66,23 +66,25 @@ def load_multiple(fn: str) -> Generator[MsmtTup, None, None]:
                         msm = ujson.loads(line)
-                        msmt_uid = trivial_id(msm)
+                        msmt_uid = trivial_id(line, msm)
+                        msm["measurement_uid"] = msmt_uid
```
Ok cool, so we are adding the `measurement_uid` to the output JSON entries. 

We should remember to update ooni/spec to reflect this improvement to the output data format.

FedericoCeratto replied:
This was based on a discussion on Slack. Now legacy msmts have the UID in them.

hellais commented on af/fastpath/fastpath/db.py:
```diff
@@ -52,14 +53,16 @@ def setup(conf) -> None:
-        input_ = ":".join(input_)
+        domain = input_[0]  # type: str
+        input_ = ",".join(input_)
+        input_ = "{" + input_ + "}"
```
Is there a reason to change the format of the input for the meek test results?

If we do implement this change, we should test if explorer is able to handle it properly. Otherwise we can maybe stick to the old format of separating the fronts with a `:`?

FedericoCeratto replied:
It's what we already have in the database.

hellais replied:
Oh, I see now that we have in the database a mixture of both formats:
```
SELECT input
FROM fastpath
WHERE (test_name = 'meek_fronted_requests_test') AND (input LIKE '{%')
LIMIT 1

┌─input────────────────────────────────────────────┐
│ {a0.awsstatic.com,d2cly7j4zqgua7.cloudfront.net} │
└──────────────────────────────────────────────────┘
```

and
```
SELECT input
FROM fastpath
WHERE (test_name = 'meek_fronted_requests_test') AND (input NOT LIKE '{%')
LIMIT 1

┌─input──────────────────────────────────────────┐
│ a0.awsstatic.com:d2zfqthxsdq309.cloudfront.net │
└────────────────────────────────────────────────┘
```

hellais replied:
So I tested the ones prefixed with `{` in explorer and while they don't look very pretty, they don't seem to break.

<img width="1379" alt="Screenshot 2022-03-03 at 18 15 38" src="https://user-images.githubusercontent.com/424620/156616446-caf24f53-fd92-44df-bc45-e07626826228.png">

I guess it's not super high priority to align these to the same format since it's a legacy test.

hellais commented on af/fastpath/fastpath/reprocessor.py:
```diff
@@ -12,12 +12,14 @@
 export AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
 PYTHONPATH=. ./fastpath/reprocessor.py ooni-data ooni-data-eu-fra-test --day 2015-1-1
 
+Note: the bundling of measurements into jsonl gz files has to remain deterministic
```
It might help to clarify more precisely what is meant by "deterministic" and why we need it.

Do we mean that the same measurements have to end up inside of the same filenames? That they end up in the same s3 paths? Or that the hash of the jsonl.gz files has to be exactly the same across runs?

FedericoCeratto replied:
It's what we discussed in the meetings and on Slack. Two runs of the reprocessor on different hosts or at different times will generate the same jsonl files.

Commit:
Add metrics and debugging

Commit:
Add domain to fastpath (currently unused on FSN)

hellais commented on af/fastpath/fastpath/reprocessor.py:
```diff
@@ -163,55 +199,71 @@ def score_measurement_and_upsert_fastpath(msm, msmt_uid, do_update: bool) -> Non
+        cc = msm.get("probe_cc").upper()
+        desc = f"{msmt_uid} {tn} {cc} {rid} {input_}"
+    except Exception as e:
+        log.info(f"Ignoring broken measurement {desc}")
```
Isn't the `desc` variable unbound in the exception handler?

FedericoCeratto replied:
Fixed.

Commit:
Add S3 update flags

Commit:
Add hashed value to JSONL file names

Commit:
Insert into fastpath on clickhouse

Commit:
Ignore malformed YAML, improve error reporting

Commit:
Handle parsing error

Commit:
Bugfix

Commit:
Improve error handling in fastpath

Commit:
Ignoring measurement with invalid measurement_start_time

hellais commented on af/fastpath/fastpath/core.py:
```diff
@@ -377,6 +377,25 @@ def match_fingerprints(measurement):
+    return default if v is None else v
+
+
+def g(d, *keys, default=None):
```
Maybe you should add a check to ensure len(keys) > 0, otherwise you will get an IndexError.

FedericoCeratto replied:
If we forget to add  the keys throwing an IndexError would be ok.

Commit:
Skip broken YAML entries without raising

Commit:
Normalize website_probe tests

Commit:
Fix meek scoring

Commit:
Handle base64 encoded header

Commit:
Handle web_conn msmt with missing body key

Commit:
Handle experimental Tor test

Commit:
Handle dash test without test_keys

Commit:
Debug json serialization failure

Commit:
Rename variable

Commit:
Fix http_header_field_manipulation, add tests

Minor changes

bassosimone commented on af/fastpath/fastpath/tests/data/http_header_field_manipulation_2.json:
```diff
@@ -0,0 +1 @@
+{"test_keys":{"requests":[],"agent":"agent","socksproxy":null},"test_start_time":"2019-10-28 11:56:47","input_hashes":[],"probe_ip":"127.0.0.1","id":"3282669b-2397-4d8f-bad3-9cd81825518e","test_helpers":{"backend":"http:\/\/37.218.247.95:80"},"probe_cc":"BR","test_runtime":1.3552720547,"input":null,"probe_city":null,"probe_asn":"AS28573","annotations":{"platform":"linux"},"software_name":"oonipro ...
```
can you indent this measurement so the diff review is easier?

FedericoCeratto replied:
Done

bassosimone commented on af/fastpath/fastpath/tests/data/http_header_field_manipulation_3.json:
```diff
@@ -0,0 +1 @@
+{"test_keys":{"requests":[{"failure":"connection_refused_error","request":{"body":null,"headers":{"aCcEPt-lANGuAGE":"en-US,en;q=0.8","AccEpt-EncodinG":"gzip,deflate,sdch","AcCePt":"text\/html,application\/xhtml+xml,application\/xml;q=0.9,*\/*;q=0.8","UseR-AgENt":"Mozilla\/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.9.2) Gecko\/20100115 Firefox\/3.6","acCePT-cHaRSEt":"ISO-8859-1,utf-8;q=0.7,*;q=0 ...
```
can you indent this measurement so the diff review is easier?

FedericoCeratto replied:
Done

Commit:
Indent data

bassosimone reviewed (approved):
🐳

Commit:
Normalize bytes bodies in YAML msmts

Commit:
Handle binary probe_city

Commit:
Skip invalid legacy JSON msmts

Example:
20160724T181310Z-GF-AS3215-web_connectivity-20160724T181226Z_AS3215_tdijfX9oJrUKsq3grDZxsFLHIcSt3ikphVa03p8b19UGgp5DDj-0.2.0-probe.json.lz4
line 980

Commit:
Skip invalid JSON

Example:
20170702T053645Z_AS204074_2ahNzbfWPPVyzqr4yxDuyp2ptVuy7BcHWH15uHPnzgoX175A3X

Commit:
Discard msmts without report_id

Commit:
Remove None as default to get()

Commit:
Discard msmts without report_id in reprocessor

Commit:
Change header decoding

Commit:
Remove "Telegam"

Commit:
Add Telegram test file

Commit:
Handle broken Telegram tests

Commit:
Normalize yaml http_invalid_request_line

Commit:
Handle incorrect padding

FedericoCeratto commented:
0.57~pr380-122 has been deployed on FSN on 2022-05-28
